### PR TITLE
fix(eslint-plugin): [space-infix-ops] fix no error when right type is function

### LINKED ISSUE #4802: [space-infix-ops] no error when right type is function
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```js
module.exports = {
	"parser": "@typescript-eslint/parser",
	"plugins": ["@typescript-eslint"],
	"extends": [
		"eslint:recommended",
		"plugin:@typescript-eslint/eslint-recommended",
		"plugin:@typescript-eslint/recommended"
	],
	"rules": {
		"@typescript-eslint/space-infix-ops": "error"
	}
}
```

```TS
// Does not throw error 👎
export interface Test {
  default?: boolean|(() => boolean);
}
```

```TS
// Does not throw error 👎
export interface Test {
  default?: (() => boolean)|(() => string);
}
```

```TS
// Correctly throws error 👍
export interface Test {
  default?: (() => boolean)|boolean;
}
```

```TS
// Correctly throws error 👍
export interface Test {
  default?: boolean | (() => boolean|Promise<boolean>);
}
```

```TS
// Correctly throws error 👍
export interface Test {
  default?: boolean|string;
}
```

**Expected Result**

I expect an error when running `npx eslint repro.ts` on the first two cases above:

```
/home/andrewda/Development/repro.ts
  2:20  error  Operator '|' must be spaced  @typescript-eslint/space-infix-ops
```

**Actual Result**

No error messages are printed by eslint when the right type is a function, e.g. `(() => string)`. If the left type is a function, errors are thrown correctly.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.18.0` |
| `@typescript-eslint/parser`        | `5.18.0` |
| `TypeScript`                       | `4.6.3` |
| `ESLint`                           | `8.13.0` |
| `node`                             | `16.13.0` |

### PR DESCRIPTION
**Pull request by tapanprakasht**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #4802
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

No errors were thrown when the right type is a function when no space between union or intersection operator. Function types are wrapped inside parenthesis. If the right type is a function, instead of getting the operator, it was always getting the wrapper parenthesis. The fix provided is to skips wrapped parenthesis for function type, so that
an operator will be properly identified.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/space-infix-ops.test.ts [hunk 1]
```diff
@@ -278,332 +384,763 @@ ruleTester.run('space-infix-ops', rule, {
...
+        }
+      `,
+      errors: [
+        {
+          messageId: 'missingSpace',
+          column: 30,
+          line: 3,
+        },
+      ],
+    },
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/space-infix-ops.test.ts:780]: Could you add a few test cases for functions that are wrapped multiple times? `((() => void))` and the like? Parenthesis and wrapping expressions are tricky.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Looks great, thanks @tapanprakasht! Just one request around added tests for edge cases and we should be good to go! 🚀

File: packages/eslint-plugin/tests/rules/space-infix-ops.test.ts (hunk 1 above)
Comment (tapanprakasht) [reply packages/eslint-plugin/tests/rules/space-infix-ops.test.ts:780]: That are nice edge cases. The current fix didn't consider this scenario with function type having multiple parenthesis wrapping. I think I need to revisit the fix again with this scenario in mind.

Comment (tapanprakasht) [reply packages/eslint-plugin/tests/rules/space-infix-ops.test.ts:1131]: @JoshuaKGoldberg Handled edge cases you have pointed out in the latest commit. Added more test cases to check the same edge cases.

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great, thanks @tapanprakasht! 🙌

**Pull request merged**